### ISSUE TITLE
Alert Notification still isn't consistent with Status

### ISSUE THREAD

**Issue by kylebrandt** [bosun-monitor/bosun#721]
https://github.com/bosun-monitor/bosun/commit/824f552763e9acd8359c47af55d0f6d6eb8c7a09 doesn't entirely fix our consistency issues. I got the following alert:

![image](https://cloud.githubusercontent.com/assets/1692624/6214328/aaf38494-b5c5-11e4-9faf-aa1c56e12e14.png)

The state is critical, but there is nothing wrong according to the contents of the email. I believe what caused this was that the alert body is only rendered after all the alerts have run:

 ```
// Check evaluates all critical and warning alert rules. An error is returned if
// the check could not be performed.
func (s *Schedule) Check(T miniprofiler.Timer, now time.Time) (time.Duration, error) {
	select {
	case s.checkRunning <- true:
		// Good, we've got the lock.
	default:
		return 0, fmt.Errorf("check already running")
	}
	r := s.NewRunHistory(now)
	start := time.Now()
	for _, a := range s.Conf.Alerts {
		s.CheckAlert(T, r, a)
	}
	d := time.Since(start)
	s.RunHistory(r)
	<-s.checkRunning
	return d, nil
}
```

Since `s.RunHistory(r)` is what builds the templates. it is run after the loop above populates the run history. 

The problem is that running all the alert can take some time. As we can see in the following log:

Bosun Log:
```
2015/02/16 04:06:25 starting check
2015/02/16 04:06:25 check alert linux.bonding start
2015/02/16 04:06:25 check alert linux.bonding done (56.517836ms): 0 crits, 0 warns
2015/02/16 04:06:25 check alert calculon.exceptions start
2015/02/16 04:06:25 check alert calculon.exceptions done (2.37762ms): 0 crits, 0 warns
2015/02/16 04:06:25 check alert haproxy.anomalous.sessions start
2015/02/16 04:06:25 check alert haproxy.anomalous.sessions done (185.584623ms): 0 crits, 0 warns
2015/02/16 04:06:25 check alert railgun_down start
2015/02/16 04:06:25 check alert railgun_down done (3.361414ms): 0 crits, 0 warns
2015/02/16 04:06:25 check alert linux.swapping start
2015/02/16 04:06:26 check alert linux.swapping done (234.911794ms): 0 crits, 0 warns
2015/02/16 04:06:26 check alert linux.entropy start
2015/02/16 04:06:26 check alert linux.entropy done (96.048156ms): 0 crits, 0 warns
2015/02/16 04:06:26 check alert linux.disk.time_per_write.95th start
2015/02/16 04:06:26 check alert linux.disk.time_per_write.95th done (265.378968ms): 0 crits, 0 warns
2015/02/16 04:06:26 check alert exceptions start
2015/02/16 04:06:26 check alert exceptions done (12.71384ms): 0 crits, 0 warns
2015/02/16 04:06:26 check alert os.net.drop_err start
2015/02/16 04:06:31 check alert os.net.drop_err done (5.061192916s): 0 crits, 0 warns
2015/02/16 04:06:31 check alert puppet.left.disabled start
2015/02/16 04:06:31 check alert puppet.left.disabled done (296.356183ms): 0 crits, 1 warns
2015/02/16 04:06:31 check alert tsd.close_wait start
2015/02/16 04:06:31 check alert tsd.close_wait done (1.665576ms): 0 crits, 0 warns
2015/02/16 04:06:31 check alert os.diskspace start
2015/02/16 04:06:37 wrote tagk: 80.47KB
2015/02/16 04:06:38 wrote tagv: 3.70MB
2015/02/16 04:06:46 wrote metrictags: 107.01MB
2015/02/16 04:06:46 wrote notifications: 54.00B
2015/02/16 04:06:46 wrote silence: 229.00B
2015/02/16 04:06:54 wrote status: 306.17MB
2015/02/16 04:06:55 wrote metadata: 18.09MB
2015/02/16 04:06:56 wrote metric: 6.02MB
2015/02/16 04:06:56 save to db complete
2015/02/16 04:07:08 wrote tagv: 3.70MB
2015/02/16 04:07:16 wrote metrictags: 107.01MB
2015/02/16 04:07:16 wrote notifications: 54.00B
2015/02/16 04:07:16 wrote silence: 229.00B
2015/02/16 04:07:25 wrote status: 306.17MB
2015/02/16 04:07:26 wrote metadata: 18.09MB
2015/02/16 04:07:26 wrote metric: 6.02MB
2015/02/16 04:07:26 wrote tagk: 80.47KB
2015/02/16 04:07:26 save to db complete
2015/02/16 04:07:39 checkUnknown
2015/02/16 04:07:45 wrote metrictags: 107.01MB
2015/02/16 04:07:45 wrote notifications: 54.00B
2015/02/16 04:07:45 wrote silence: 229.00B
2015/02/16 04:07:48 check alert os.diskspace done (1m16.507231634s): 0 crits, 3 warns
2015/02/16 04:07:48 check alert linux.disk.time_per_read.95th start
2015/02/16 04:07:48 check alert linux.disk.time_per_read.95th done (238.628474ms): 0 crits, 0 warns
2015/02/16 04:07:48 check alert ntp.high_offset start
2015/02/16 04:07:49 check alert ntp.high_offset done (710.563274ms): 0 crits, 0 warns
2015/02/16 04:07:49 check alert anomalous.temperature start
2015/02/16 04:07:49 check alert anomalous.temperature done (39.611302ms): 0 crits, 0 warns
2015/02/16 04:07:49 check alert lb.ip_count_changed start
2015/02/16 04:07:49 check alert lb.ip_count_changed done (9.897482ms): 0 crits, 0 warns
2015/02/16 04:07:49 check alert linux.high_cpu_per_core start
2015/02/16 04:07:50 check alert linux.high_cpu_per_core done (984.053145ms): 0 crits, 0 warns
2015/02/16 04:07:50 check alert puppet.failed_resources start
2015/02/16 04:07:50 check alert puppet.failed_resources done (22.891113ms): 0 crits, 2 warns
2015/02/16 04:07:50 check alert asterisk.office_numbers.nyhq_main_line start
2015/02/16 04:07:50 check alert asterisk.office_numbers.nyhq_main_line done (2.427395ms): 0 crits, 0 warns
2015/02/16 04:07:50 check alert it.site.down start
2015/02/16 04:07:50 check alert it.site.down done (21.885825ms): 0 crits, 0 warns
2015/02/16 04:07:50 check alert vrtx.state.chassis start
2015/02/16 04:07:50 check alert vrtx.state.chassis done (2.653255ms): 0 crits, 0 warns
2015/02/16 04:07:50 check alert scollector.log.errors start
2015/02/16 04:07:50 check alert scollector.log.errors done (441.504212ms): 0 crits, 0 warns
2015/02/16 04:07:50 check alert linux.net.conntrack.percent.used start
2015/02/16 04:07:50 check alert linux.net.conntrack.percent.used done (32.764283ms): 0 crits, 0 warns
2015/02/16 04:07:50 check alert asterisk.call.quality start
2015/02/16 04:07:50 check alert asterisk.call.quality done (2.119034ms): 0 crits, 0 warns
2015/02/16 04:07:50 check alert haproxylogs.server_err.percent start
2015/02/16 04:07:50 check alert haproxylogs.server_err.percent done (5.074609ms): 0 crits, 0 warns
2015/02/16 04:07:50 check alert opentsdb.relay start
2015/02/16 04:07:50 check alert opentsdb.relay done (1.085193ms): 0 crits, 0 warns
2015/02/16 04:07:50 check alert haproxy_session_limit start
2015/02/16 04:07:51 check alert haproxy_session_limit done (180.148105ms): 0 crits, 0 warns
2015/02/16 04:07:51 check alert linux.tcp start
2015/02/16 04:07:51 check alert linux.tcp done (630.985088ms): 0 crits, 1 warns
2015/02/16 04:07:51 check alert hardware start
2015/02/16 04:07:51 check alert hardware done (8.085961ms): 0 crits, 3 warns
2015/02/16 04:07:51 check alert vrtx.state.power start
2015/02/16 04:07:51 check alert vrtx.state.power done (2.765963ms): 0 crits, 0 warns
2015/02/16 04:07:51 check alert vrtx.chassis.disk start
2015/02/16 04:07:51 check alert vrtx.chassis.disk done (52.510381ms): 0 crits, 0 warns
2015/02/16 04:07:51 check alert logstash.docs start
2015/02/16 04:07:52 check alert logstash.docs done (426.799595ms): 0 crits, 0 warns
2015/02/16 04:07:52 check alert redis.replication start
2015/02/16 04:07:52 check alert redis.replication done (60.233887ms): 0 crits, 0 warns
2015/02/16 04:07:52 check alert logstash.vrrp.check start
2015/02/16 04:07:52 check alert logstash.vrrp.check done (569.049678ms): 0 crits, 0 warns
2015/02/16 04:07:52 check alert it.vrtx.anomalous_load start
2015/02/16 04:07:52 check alert it.vrtx.anomalous_load done (61.119098ms): 0 crits, 0 warns
2015/02/16 04:07:52 check alert calculon.providence.calls start
2015/02/16 04:07:52 check alert calculon.providence.calls done (9.654243ms): 0 crits, 0 warns
2015/02/16 04:07:52 check alert asterisk.office_numbers.uk_test_number start
2015/02/16 04:07:52 check alert asterisk.office_numbers.uk_test_number done (1.905085ms): 0 crits, 0 warns
2015/02/16 04:07:52 check alert elastic.cluster.status start
2015/02/16 04:07:52 check alert elastic.cluster.status done (5.582461ms): 0 crits, 0 warns
2015/02/16 04:07:52 check alert netbackup start
2015/02/16 04:07:53 check alert netbackup done (58.527708ms): 0 crits, 1 warns
2015/02/16 04:07:53 check alert redis.fragmentation start
2015/02/16 04:07:53 check alert redis.fragmentation done (196.201087ms): 0 crits, 2 warns
2015/02/16 04:07:53 check alert it.ping.packetloss start
2015/02/16 04:07:53 check alert it.ping.packetloss done (42.813077ms): 0 crits, 0 warns
2015/02/16 04:07:53 check alert or_hosts_down start
2015/02/16 04:07:53 check alert or_hosts_down done (47.799248ms): 0 crits, 0 warns
2015/02/16 04:07:53 check alert mysql.slow_queries start
2015/02/16 04:07:53 check alert mysql.slow_queries done (3.84645ms): 0 crits, 0 warns
2015/02/16 04:07:53 check alert haproxy.or.websockets start
2015/02/16 04:07:53 check alert haproxy.or.websockets done (114.577544ms): 0 crits, 0 warns
2015/02/16 04:07:53 check alert it.logstash.output.rate start
2015/02/16 04:07:53 check alert it.logstash.output.rate done (2.562302ms): 1 crits, 0 warns
2015/02/16 04:07:53 check alert asterisk.office_numbers.usa_toll_free start
2015/02/16 04:07:53 check alert asterisk.office_numbers.usa_toll_free done (2.214965ms): 0 crits, 0 warns
2015/02/16 04:07:53 check alert hbase.replication_queue start
2015/02/16 04:07:53 check alert hbase.replication_queue done (3.26322ms): 0 crits, 0 warns
2015/02/16 04:07:53 check alert slower.route.performance start
2015/02/16 04:07:53 wrote status: 306.17MB
2015/02/16 04:07:55 wrote metadata: 18.09MB
2015/02/16 04:07:55 wrote metric: 6.02MB
2015/02/16 04:07:55 wrote tagk: 80.47KB
2015/02/16 04:07:56 wrote tagv: 3.70MB
2015/02/16 04:07:56 save to db complete
2015/02/16 04:08:01 check alert slower.route.performance done (7.847553289s): 0 crits, 0 warns
2015/02/16 04:08:01 check alert bosun_build start
2015/02/16 04:08:01 check alert bosun_build done (4.743103ms): 0 crits, 0 warns
2015/02/16 04:08:01 check alert offsite_backup start
2015/02/16 04:08:01 check alert offsite_backup done (3.92396ms): 0 crits, 0 warns
2015/02/16 04:08:01 check alert high_cpu start
2015/02/16 04:08:03 check alert high_cpu done (1.873586842s): 1 crits, 4 warns
2015/02/16 04:08:03 check alert redis.save.status start
2015/02/16 04:08:03 check alert redis.save.status done (106.120055ms): 0 crits, 0 warns
2015/02/16 04:08:03 check alert haproxy.server.error_connections start
2015/02/16 04:08:03 check alert haproxy.server.error_connections done (646.333711ms): 0 crits, 1 warns
2015/02/16 04:08:03 check alert haproxy.server.downtime start
2015/02/16 04:08:05 check alert haproxy.server.downtime done (1.713828097s): 1 crits, 0 warns
2015/02/16 04:08:05 check alert puppet.disabled.no_reason start
2015/02/16 04:08:05 check alert puppet.disabled.no_reason done (83.070825ms): 0 crits, 0 warns
2015/02/16 04:08:05 check alert asterisk.office_numbers.denver_main_line start
2015/02/16 04:08:05 check alert asterisk.office_numbers.denver_main_line done (2.528141ms): 0 crits, 0 warns
2015/02/16 04:08:05 check alert it.java.heap.alert start
2015/02/16 04:08:05 check alert it.java.heap.alert done (36.027944ms): 0 crits, 0 warns
2015/02/16 04:08:05 check alert os.low.memory start
2015/02/16 04:08:05 check alert os.low.memory done (240.487872ms): 0 crits, 0 warns
2015/02/16 04:08:07 wrote tagk: 80.47KB
2015/02/16 04:08:08 wrote tagv: 3.70MB
2015/02/16 04:08:16 wrote metrictags: 107.01MB
2015/02/16 04:08:16 wrote notifications: 54.00B
2015/02/16 04:08:16 wrote silence: 229.00B
2015/02/16 04:08:21 check took 1m40.27880697s
2015/02/16 04:08:21 relayed alert linux.tcp{host=ny-bosun02} to [sre-alerts@stackoverflow.com] sucessfully
2015/02/16 04:08:21 relayed alert haproxy.server.error_connections{} to [sre-alerts@stackoverflow.com] sucessfully
2015/02/16 04:08:21 relayed alert haproxy.server.downtime{} to [sre-alerts@stackoverflow.com] sucessfully
2015/02/16 04:08:24 wrote status: 307.31MB
2015/02/16 04:08:25 wrote metadata: 18.09MB
2015/02/16 04:08:26 wrote metric: 6.02MB
2015/02/16 04:08:26 save to db complete
2015/02/16 04:08:37 wrote silence: 229.00B
2015/02/16 04:08:45 wrote status: 307.30MB
2015/02/16 04:08:46 wrote metadata: 18.09MB
2015/02/16 04:08:47 wrote metric: 6.02MB
2015/02/16 04:08:47 wrote tagk: 80.47KB
2015/02/16 04:08:47 wrote tagv: 3.70MB
2015/02/16 04:08:54 checkUnknown
2015/02/16 04:08:55 wrote metrictags: 107.01MB
2015/02/16 04:08:55 wrote notifications: 54.00B
2015/02/16 04:08:55 save to db complete
2015/02/16 04:09:08 wrote metric: 6.02MB
2015/02/16 04:09:08 wrote tagk: 80.47KB
2015/02/16 04:09:08 wrote tagv: 3.70MB
2015/02/16 04:09:16 wrote metrictags: 107.01MB
2015/02/16 04:09:16 wrote notifications: 54.00B
2015/02/16 04:09:16 wrote silence: 229.00B
```

When this is combined with a delay in scollector queueing data the information isn't consistent between state evaluation and rendering of the notification. We can see this happened in this case:

![image](https://cloud.githubusercontent.com/assets/1692624/6214447/9e8bf032-b5c6-11e4-994e-c388fbf42e41.png)

However using the rule page to look back, things seems fine with the timeline because the data is all filled in now:

![image](https://cloud.githubusercontent.com/assets/1692624/6214503/042b1292-b5c7-11e4-8ab8-9ee6c16aa13c.png)

Since we were also writing to the database during these checks that might have contributed to ny-lb05 queueing data. So possible ideas:

 1. Make the template rendering happen at the same as the alert check runs
 2. Don't save the db while the alert check is running? (Sounds like it could be dangerous)

**Comment by Dieterbe** [comment]
this is probably the same problem as #700 ? Also we shouldn't solely rely on timing because that will never guarantee consistency.  I believe the proper fix will be #715 which guarantees that, within 1 alerting run, any query for the same data, will yield the same result.  it's kindof as if the entire alerting run works on a snapshot.

**Comment by kylebrandt** [comment]
I feel cache is the right way to fix it as well. They gives us the best possible consistency. What @mjibson wants to try first is make a global "lag" for now so alerts are running mostly after it has all the data.

I the lag idea combined with making template eval happen as close as possible in time to deciding the state would make this far less frequent. But passing a cache seems like the reliable way to be consistent me.

**Comment by Dieterbe** [comment]
i just realized cache size would have to be unlimited otherwise you run the risk of querying the same data twice. and an unlimited cache size is a pretty sloppy way to merely rule out the issue where you query twice you get different data, because it means caching all other data as well, specifically data that will only be read once (during the alerting run) and which potentially takes up a lot of space. perhaps we should do something more clever, like first collect all queries and then only cache the ones that are repeat queries.  but this might make things more complicated than needed.  i don't see a cache that has an unbound size but only lasts for the alerting run as a threatening problem. (because you're not supposed to query that much unique data that this becomes a problem)

**Comment by Dieterbe** [comment]
cache is now merged (#715) with infinite size (on a per alerting run basis). that should rule this out.

**Comment by mjibson** [comment]
Reopen if this happens again.
